cryptosuite/bbsplussignaturesuite.go

```diff
@@ -0,0 +1,355 @@
+	if err == nil {
+		return signatureBytes, nil
+	}
+	return nil, errors.New("unknown encoding of proof value")
+}
+
+// CryptoSuiteProofType interface
+
+func (BBSPlusSignatureSuite) Marshal(data interface{}) ([]byte, error) {
+	// JSONify the provable object
+	jsonBytes, err := json.Marshal(data)
+	if err != nil {
+		return nil, err
+	}
+	return jsonBytes, nil
+}
+
+func (BBSPlusSignatureSuite) Canonicalize(marshaled []byte) (*string, error) {
+	// the LD library anticipates a generic golang json object to normalize
+	var generic map[string]interface{}
+	if err := json.Unmarshal(marshaled, &generic); err != nil {
+		return nil, err
+	}
+	normalized, err := LDNormalize(generic)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not canonicalize provable document")
+	}
+	canonicalString := normalized.(string)
+	return &canonicalString, nil
+}
+
+// CreateVerifyHash https://w3c-ccg.github.io/data-integrity-spec/#create-verify-hash-algorithm
+// augmented by https://w3c-ccg.github.io/ldp-bbs2020/#create-verify-data-algorithm
+func (b BBSPlusSignatureSuite) CreateVerifyHash(provable Provable, proof crypto.Proof, opts *ProofOptions) ([]byte, error) {
+	// first, make sure "created" exists in the proof and insert an LD context property for the proof vocabulary
+	preparedProof, err := b.prepareProof(proof, opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not prepare proof for the create verify hash algorithm")
+	}
+
+	// marshal provable to prepare for canonicalizaiton
+	marshaledProvable, err := b.Marshal(provable)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not marshal provable")
+	}
+
+	// canonicalize provable using the suite's method
+	canonicalProvable, err := b.Canonicalize(marshaledProvable)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not canonicalize provable")
+	}
+
+	// marshal proof to prepare for canonicalizaiton
+	marshaledOptions, err := b.Marshal(preparedProof)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not marshal proof")
+	}
+
+	// 4.1 canonicalize  proof using the suite's method
+	canonicalizedOptions, err := b.Canonicalize(marshaledOptions)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not canonicalize proof")
+	}
+
+	// 4.2 set output to the result of the hash of the canonicalized options document
+	canonicalizedOptionsBytes := []byte(*canonicalizedOptions)
+	optionsDigest, err := b.Digest(canonicalizedOptionsBytes)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not take digest of proof")
+	}
+
+	// 4.3 hash the canonicalized doc and append it to the output
+	canonicalDoc := []byte(*canonicalProvable)
+	documentDigest, err := b.Digest(canonicalDoc)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not take digest of provable")
+	}
+
+	// 5. return the output
+	output := append(optionsDigest, documentDigest...)
+	return output, nil
+}
+
+func (BBSPlusSignatureSuite) Digest(tbd []byte) ([]byte, error) {
+	// handled by the algorithm itself
```

Comment:
I think this is confusing. The [spec](https://w3c-ccg.github.io/ldp-bbs2020/#the-bbs-signature-suite-2020) says that Digest is Blake2. Why isn't that the case here?

Reply:
because the signature/verification algorithm uses blake2 we would be blake2ing a blake2 if we added this

Reply:
added a comment to the interface and updated outdated links

Reply:
Thanks. That makes sense to me now. 

The interface, as it stands, is promoting strong coupling across `CryptoSuiteProofType` and a signature algorithm (which is unclear in which interface it's defined). That makes it difficult to know what the expectations are when you're dealing with an implementation. Additionally, it's leaking implementation details because users will need to understand what implementation of the interface is being used. 

I think the root cause of this is that `CreateVerifyHash()` is part of the `CryptoSuiteProofType` interface. Instead, what if there is a separate function with a parameter of 
 type `CryptoSuiteProofType`? If you need to specify whether the signing algo does digest within it, another method like `IsSignDigested() bool` to `CryptoSuiteProofType`. WDYT?

Reply:
> the strong coupling is necessary because a suite defines a proof type

Coupling should never be necessary =)

👍  to separate this into another issue.